pkg/service/jetstream.go

```diff
@@ -0,0 +1,212 @@
+		b.streamSubjects[s] = len(b.streams) - 1
+	}
+
+	return nil
+}
+
+// RemoveStream will attempt to remove consumers and streams based on a list of subjects.
+// List of subjects must be exactly the same as was used in AddStream since js Stream and js Consumer names
+// are based on the subjects.
+func (b *Service) RemoveStream(subjects ...string) error {
+	if b.js == nil {
+		return ErrNotAvailable
+	}
+
+	hash := b.jsMakeHash(subjects...)
+	streamName := b.jsStreamName(hash)
+	consumerName := b.jsConsumerName(hash)
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(b.Context, 30*time.Second)
+	defer cancel()
+
+	err := b.js.DeleteConsumer(streamName, consumerName, nats.Context(ctx))
+	if err != nil {
+		return fmt.Errorf("DeleteConsumer failed: %w", err)
+	}
+	err = b.js.DeleteStream(streamName, nats.Context(ctx))
+	if err != nil {
+		return fmt.Errorf("DeleteStream failed: %w", err)
+	}
+
+	b.removeStreamsFromMap(subjects)
+
+	return nil
+}
+
+func (b *Service) removeStreamsFromMap(subjects []string) {
+	idsToDeleteSet := make(map[int]struct{}, len(subjects))
+	for _, s := range subjects {
+		if ssIdx, ok := b.streamSubjects[s]; ok {
+			delete(b.streamSubjects, s)
+			idsToDeleteSet[ssIdx] = struct{}{}
+		}
+	}
+
+	idsToDelete := make([]int, 0, len(idsToDeleteSet))
+	for id := range idsToDelete {
+		idsToDelete = append(idsToDelete, id)
+	}
+	slices.SortFunc(idsToDelete, func(a, b int) int { return b - a })
+
+	for id := range idsToDelete {
+		b.streams = slices.Delete(b.streams, id, id+1)
+	}
+}
+
+func (b *Service) attemptJSConsume(handler nats.MsgHandler, subject string) (*nats.Subscription, error) {
+	if b.js == nil {
+		return nil, ErrNotAvailable
+	}
+
+	var info jsStream
+	streamName, consumerName := "", ""
+	if id, ok := b.streamSubjects[subject]; !ok {
+		return nil, ErrNotAvailable
+	} else {
+		info = b.streams[id]
+		streamName = info.cfgStream.Name
+		consumerName = info.consumerInfo.Config.Durable
+	}
+
+	sub, err := b.js.PullSubscribe(streamName, consumerName, nats.ManualAck(), nats.Bind(streamName, consumerName))
+	if err != nil {
+		return nil, err
+	}
+
+	b.Group.Go(func() error {
+		log.Println("PullSubscribe loop start: ", subject)
+		defer log.Println("PullSubscribe loop exit: ", subject)
+		for {
+			select {
+			case <-b.Context.Done():
+				return b.Context.Err()
+			default:
+			}
+
+			ctx, cancel := context.WithTimeout(b.Context, 5*time.Second)
+			defer cancel()
+
+			msgs, err := sub.Fetch(10, nats.Context(ctx))
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					return fmt.Errorf("context cancelled during pulling next message: %w", err)
+				}
+				if errors.Is(err, context.DeadlineExceeded) {
+					continue
+				}
+				if errors.Is(err, nats.ErrBadSubscription) {
+					log.Printf("subscription to %s closed", subject)
+					return nil
+				}
+				return fmt.Errorf("pulling message failed: %w", err)
+			}
+			for _, msg := range msgs {
+				handler(msg)
```

Comment:
I, as a user, would expect to have a possibility to return error from a handler, so the message won't be Acked, if there is an error.
I see you are trying to use the same handler as for normal subscription, so I don't know if you  are going to re-implement it. So it is just my wish, probably :)

Reply:
Maybe in the next iteration. In order to return an error from the handler, the handler signature has to be changed. I don't want to break compatibility with existing users just yet :)